camel/functions/retrieve_tools_function.py

```diff
@@ -0,0 +1,81 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from typing import List, Optional
+
+import numpy as np
+import torch
+from transformers import AutoModel, AutoTokenizer
+
+from camel.functions.openai_function import OpenAIFunction
+
+
+def embed_text(text: Optional[str], model, tokenizer):
+    inputs = tokenizer(text, return_tensors="pt")
+    with torch.no_grad():
+        outputs = model(**inputs)
+
+    return outputs.last_hidden_state.mean(dim=1).squeeze().cpu().numpy()
+
+
+def cosine_similarity(vec1: np.ndarray, vec2: np.ndarray) -> float:
+    return np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))
+
+
+def retrieve_tools(
+    query: str, tools: List[OpenAIFunction], k: int = 3
+) -> Optional[List[OpenAIFunction]]:
+    r"""Using semantic search to retrieve relevent tools
+    OpenAI / Open source Embedding models to embed the query and the schemas / docstring of `OpenAIFunction`
+    Find the most relevent ones based on their embedding vectors
+
+    Args:
+        tools (List): The tools to be searched.
+
+    Returns:
+        retrieved_tools (List): List of retrieved (most relevant) tools.
+    """
+    tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+    model = AutoModel.from_pretrained("bert-base-uncased")
+
+    query_embedding = embed_text(query, model, tokenizer)
+
+    similarities = []
+    for func in tools:
+        func_embedding = embed_text(func.func.__doc__, model, tokenizer)
```

Comment:
add handling for cases when the tools list is empty or all provided descriptions are `None`?

Reply:
hasn't updated